lighthouse-core/test/gather/mock-driver.js

```diff
@@ -0,0 +1,130 @@
+    if (delay) return new Promise(resolve => setTimeout(() => resolve(response), delay));
+    return Promise.resolve(response);
+  });
+
+  mockFn.mockResponse = (command, response, delay) => {
+    mockResponses.push({command, response, delay});
+    return mockFn;
+  };
+
+  mockFn.findInvocation = command => {
+    expect(mockFn).toHaveBeenCalledWith(command, expect.anything());
+    return mockFn.mock.calls.find(call => call[0] === command)[1];
+  };
+
+  return mockFn;
+}
+
+/**
+ * Creates a jest mock function whose implementation invokes `.on`/`.once` listeners after a setTimeout tick.
+ * Closely mirrors `createMockSendCommandFn`.
+ *
+ * It is decorated with two methods:
+ *    - `mockEvent` which pushes protocol event payload for consumption
+ *    - `findListener` which asserts that `on` was invoked with the given event name and
+ *      returns the listener .
+ */
+function createMockOnceFn() {
+  const mockEvents = [];
+  const mockFn = jest.fn().mockImplementation((eventName, listener) => {
+    const indexOfResponse = mockEvents.findIndex(entry => entry.event === eventName);
+    if (indexOfResponse === -1) return;
+    const {response} = mockEvents[indexOfResponse];
+    mockEvents.splice(indexOfResponse, 1);
+    // Wait a tick because real events never fire immediately
+    setTimeout(() => listener(response), 0);
+  });
+
+  mockFn.mockEvent = (event, response) => {
+    mockEvents.push({event, response});
+    return mockFn;
+  };
+
+  mockFn.findListener = event => {
+    expect(mockFn).toHaveBeenCalledWith(event, expect.anything());
+    return mockFn.mock.calls.find(call => call[0] === event)[1];
+  };
+
+  return mockFn;
+}
+
+/**
+ * Very much like `createMockOnceFn`, but will fire all the events (not just one for every call).
+ * So it's good for .on w/ many events.
+ */
+function createMockOnFn() {
+  const mockEvents = [];
+  const mockFn = jest.fn().mockImplementation((eventName, listener) => {
+    const events = mockEvents.filter(entry => entry.event === eventName);
+    if (!events.length) return;
+    for (const event of events) {
+      const indexOfEvent = mockEvents.indexOf(event);
+      mockEvents.splice(indexOfEvent, 1);
+    }
+    // Wait a tick because real events never fire immediately
+    setTimeout(() => {
+      for (const event of events) {
+        listener(event.response);
+      }
+    }, 0);
+  });
+
+  mockFn.mockEvent = (event, response) => {
+    mockEvents.push({event, response});
+    return mockFn;
+  };
+
+  mockFn.findListener = event => {
+    expect(mockFn).toHaveBeenCalledWith(event, expect.anything());
+    return mockFn.mock.calls.find(call => call[0] === event)[1];
+  };
+
+  return mockFn;
+}
+
+/**
+ * In some functions we have lots of promise follow ups that get queued by protocol messages.
+ * This is a convenience method to easily advance all timers and flush all the queued microtasks.
+ */
+async function flushAllTimersAndMicrotasks() {
```

Comment:
is this definitely necessary for `source-maps-test.js`? I believe @patrickhulce and I talked way back when about how we'd be able to not rely on this magic (or at least automate it away) but it was never a high priority. If it could not be needed outside `driver-test.js`, though, that would be ideal.

Reply:
> is this definitely necessary for source-maps-test.js? I believe @patrickhulce and I talked way back when about how we'd be able to not rely on this magic (or at least automate it away) but it was never a high priority. If it could not be needed outside driver-test.js, though, that would be ideal.

looks like the answer is "sort of", but just a line of `jest.advanceTimersByTime(1);` is sufficient for `source-maps-test.js`, it doesn't need the full `flushAllTimersAndMicrotasks()`